### PR TITLE
feat: add gc for the main chain

### PR DESCRIPTION
**Pull request by raventid**

### CODE REVIEW

File: contract/src/lib.rs [hunk 1]
```diff
@@ -388,6 +397,48 @@ impl Contract {
             false
         }
     }
+
+    // Current GC implementation is doing full travers of the blocks starting from the main chain
+    // tip. We can optimize this, by storing the height of a current genesis block in our state.
+    #[allow(dead_code)]
+    fn run_gc(&mut self) {
```
[the remark this answers is not in the document]
Comment (raventid) [reply contract/src/lib.rs]: added a param

File: contract/src/lib.rs [hunk 2]
```diff
@@ -388,6 +397,48 @@ impl Contract {
             false
         }
     }
+
+    // Current GC implementation is doing full travers of the blocks starting from the main chain
+    // tip. We can optimize this, by storing the height of a current genesis block in our state.
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (raventid) [reply contract/src/lib.rs]: done

File: contract/src/lib.rs [hunk 3]
```diff
@@ -388,6 +402,50 @@ impl Contract {
...
+
+        let tip_blockheader = self
+            .headers_pool
+            .get(&self.mainchain_tip_blockhash)
+            .expect("tip blockheader must be in a header pool");
+
+        let amount_of_headers_we_store =
+            tip_blockheader.block_height - initial_blockheader.block_height;
+
+        if amount_of_headers_we_store > self.gc_threshold as u64 {
```
Comment (karim-en) [inline contract/src/lib.rs:422]: It is better to declare `gc_threshold` as u64 instead of casting it.

File: contract/src/lib.rs [hunk 4]
```diff
@@ -388,6 +402,50 @@ impl Contract {
             false
         }
     }
+
+    // Current GC implementation is doing full travers of the blocks starting from the main chain
+    // tip. We can optimize this, by storing the height of a current genesis block in our state.
+    pub fn run_mainchain_gc(&mut self, batch_size: u8) {
```
Comment (karim-en) [inline contract/src/lib.rs:408]: It is better to declare batch_size as u64 instead of casting it.

File: contract/src/lib.rs [hunk 5]
```diff
@@ -388,6 +402,50 @@ impl Contract {
...
+
+            let start_removal_height = initial_blockheader.block_height;
+            let end_removal_height = initial_blockheader.block_height + selected_amount_to_remove;
+
+            for height in start_removal_height..end_removal_height {
+                let blockhash = self
+                    .mainchain_height_to_header
+                    .get(&height)
+                    .expect("target height should exist")
+                    .to_string();
```
Comment (karim-en) [inline contract/src/lib.rs:435]: The type is already a string; you don't need `to_string` here.

File: contract/src/lib.rs [hunk 6]
```diff
@@ -388,6 +402,50 @@ impl Contract {
...
+                self.mainchain_header_to_height.remove(&blockhash);
+                self.mainchain_height_to_header.remove(&height);
+                self.headers_pool.remove(&blockhash);
+            }
+
+            let new_initial_blockhash = self
+                .mainchain_height_to_header
+                .get(&end_removal_height)
+                .expect("target height should exist");
+            self.mainchain_initial_blockhash = new_initial_blockhash.to_string();
```
Comment (karim-en) [inline contract/src/lib.rs:446]: The type is already a string; you don't need to_string here.
also, let's combine these two actions in one line.

**Review by karim-en — CHANGES_REQUESTED** [review]

File: contract/src/lib.rs (hunk 6 above)
Comment (raventid) [reply contract/src/lib.rs]: it is &String here, but I change code in a slightly different way (though still involves cloning the string)

File: contract/src/lib.rs (hunk 5 above)
Comment (raventid) [reply contract/src/lib.rs]: rewrote in a different way

File: contract/src/lib.rs (hunk 4 above)
Comment (raventid) [reply contract/src/lib.rs]: done

File: contract/src/lib.rs (hunk 3 above)
Comment (raventid) [reply contract/src/lib.rs]: done

File: contract/src/lib.rs [hunk 7]
```diff
@@ -412,6 +426,45 @@ impl Contract {
...
+
+            let start_removal_height = initial_blockheader.block_height;
+            let end_removal_height = initial_blockheader.block_height + selected_amount_to_remove;
+
+            for height in start_removal_height..end_removal_height {
+                let blockhash = self.mainchain_height_to_header[&height].clone();
+
+                self.mainchain_header_to_height.remove(&blockhash);
+                self.mainchain_height_to_header.remove(&height);
+                self.headers_pool.remove(&blockhash);
```
Comment (karim-en) [inline contract/src/lib.rs:461]:
```suggestion
                let blockhash = &self.mainchain_height_to_header[&height];

                self.headers_pool.remove(blockhash);
                self.mainchain_header_to_height.remove(blockhash);
                self.mainchain_height_to_header.remove(&height);
```

File: contract/src/lib.rs [hunk 8]
```diff
@@ -412,6 +426,45 @@ impl Contract {
...
+    /// Public call to run GC on a mainchain.
+    /// batch_size is how many block headers should be removed in the execution
+    ///
+    /// # Panics
+    /// If initial blockheader or tip blockheader are not in a header pool
+    pub fn run_mainchain_gc(&mut self, batch_size: u64) {
+        let initial_blockheader = self
+            .headers_pool
+            .get(&self.mainchain_initial_blockhash)
+            .expect("initial blockheader must be in a header pool");
```
Comment (karim-en) [inline contract/src/lib.rs:439]: Please replace it with  `.unwrap_or_else(|| env::panic_str("your error"))`

see this https://docs.near.org/build/smart-contracts/anatomy/reduce-size

File: contract/src/lib.rs [hunk 9]
```diff
@@ -412,6 +426,45 @@ impl Contract {
...
+
+        let tip_blockheader = self
+            .headers_pool
+            .get(&self.mainchain_tip_blockhash)
+            .expect("tip blockheader must be in a header pool");
```
Comment (karim-en) [inline contract/src/lib.rs:444]: Please replace it with  `.unwrap_or_else(|| env::panic_str("your error"))`

**Review by karim-en — CHANGES_REQUESTED** [review]

File: contract/src/lib.rs (hunk 9 above)
Comment (raventid) [reply contract/src/lib.rs]: Changed `expect()` to `panic_str` in https://github.com/aurora-is-near/btc-light-client-contract/pull/9/commits/2c765163713994bb6cdd22daf20eba487a1b2c68

File: contract/src/lib.rs (hunk 8 above)
Comment (raventid) [reply contract/src/lib.rs]: Changed `expect()` to `panic_str` in https://github.com/aurora-is-near/btc-light-client-contract/pull/9/commits/2c765163713994bb6cdd22daf20eba487a1b2c68

File: contract/src/lib.rs (hunk 7 above)
Comment (raventid) [reply contract/src/lib.rs]: Updated in https://github.com/aurora-is-near/btc-light-client-contract/commit/2c765163713994bb6cdd22daf20eba487a1b2c68

**Review by karim-en — APPROVED** [review]

**Pull request merged**